dali/pipeline/operators/geometric/flip.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include "dali/pipeline/operators/geometric/flip.h"
+#include "dali/kernels/imgproc/flip_cpu.h"
+#include "dali/kernels/kernel_params.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/util/ocv.h"
+#include "dali/pipeline/operators/geometric/flip_util.h"
+
+namespace dali {
+
+DALI_SCHEMA(Flip)
+    .DocStr(R"code(Flip the image on the horizontal and/or vertical axes.)code")
+    .NumInput(1)
+    .NumOutput(1)
+    .AllowMultipleInputSets()
+    .AddOptionalArg("horizontal", R"code(Perform a horizontal flip.)code", 1, true)
+    .AddOptionalArg("vertical", R"code(Perform a vertical flip.)code", 0, true);
+
+
+template <>
+Flip<CPUBackend>::Flip(const OpSpec &spec)
+    : Operator<CPUBackend>(spec) {}
+
+void RunFlip(Tensor<CPUBackend> &output, const Tensor<CPUBackend> &input,
+    bool horizontal, bool vertical) {
+  DALI_TYPE_SWITCH(
+      input.type().id(), DType,
+      auto output_ptr = output.mutable_data<DType>();
+      auto input_ptr = input.data<DType>();
+      auto kernel = kernels::FlipCPU<DType>();
+      kernels::KernelContext ctx;
+      auto shape = TransformShapes({input.shape()}, input.GetLayout() == DALI_NHWC)[0];
+      auto in_view = kernels::InTensorCPU<DType, 4>(input_ptr, shape);
```

Comment:
wondering, why 4? Isn't your input 3D ?

Reply:
The kernel works for volumetric data and in the operator the depth dimension is used for color planes in nchw layout.

Reply:
So it can be used with the sequences now? If so please add `.AllowSequences(false)` to schema.

Reply:
No it can't be used with sequences for now. Even though the kernel could be adapted for sequences, the operator would demand bigger rework.